pkg/remotecmd/kubeexec.go

```diff
@@ -0,0 +1,294 @@
+	pidWriterCmd := fmt.Sprintf("echo $$ > %s", pidFile)
+	// Redirecting to /proc/1/fd/* allows to redirect the process output to the output streams of PID 1 process inside the container.
+	// This way, returning the container logs with 'odo logs' or 'kubectl logs' would work seamlessly.
+	// See https://stackoverflow.com/questions/58716574/where-exactly-do-the-logs-of-kubernetes-pods-come-from-at-the-container-level
+	outputRedirectCmd := "1>>/proc/1/fd/1 2>>/proc/1/fd/2"
+	if devfileCmd.Exec.WorkingDir != "" {
+		// since we are using /bin/sh -c, the command needs to be within a single double quote instance,
+		// for example "cd /tmp && pwd"
+		// Full command is: /bin/sh -c "echo $$ > $pidFile && cd $workingDir && ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2"
+		// We are deleting the PID file if the main command does not succeed, so its status could be detected accordingly.
+		cmd = append(cmd, fmt.Sprintf("%s && cd %s && (%s) %s",
+			pidWriterCmd, devfileCmd.Exec.WorkingDir, cmdLine, outputRedirectCmd))
+	} else {
+		// Full command is: /bin/sh -c "echo $$ > $pidFile && ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2"
+		// We are deleting the PID file if the main command does not succeed, so its status could be detected accordingly.
+		cmd = append(cmd, fmt.Sprintf("%s && (%s) %s", pidWriterCmd, cmdLine, outputRedirectCmd))
+	}
+
+	go func() {
+		if outputHandler != nil {
+			outputHandler(Starting, nil, nil, nil)
+		}
+
+		stdout, stderr, err := ExecuteCommandAndGetOutput(kclient, podName, containerName, false, cmd...)
+		if err != nil {
+			klog.V(2).Infof("error while running background command: %v", err)
+		}
+
+		if outputHandler != nil {
+			outputHandler(Stopped, stdout, stderr, err)
+		}
+	}()
+
+	return nil
+}
+
+// StopProcessForCommand stops the process representing the specified Devfile command.
+// Because of the way this process is launched and its PID stored (see StartProcessForCommand),
+// we need to determine the process children (there should be only one child). Then killing those children
+// will exit the parent 'sh' process.
+func (k *kubeExecProcessHandler) StopProcessForCommand(
+	devfileCmd devfilev1.Command,
+	kclient kclient.ClientInterface,
+	podName string,
+	containerName string,
+) error {
+	klog.V(4).Infof("StopProcessForCommand for %q", devfileCmd.Id)
+	defer func() {
+		pidFile := getPidFileForCommand(devfileCmd)
+		err := ExecuteCommand([]string{common.ShellExecutable, "-c", fmt.Sprintf("rm -f %s", pidFile)},
+			kclient, podName, containerName, false, nil, nil)
+		if err != nil {
+			klog.V(2).Infof("Could not remove file %q: %v", pidFile, err)
+		}
+	}()
+
+	ppid, err := getRemoteProcessPID(kclient, devfileCmd, podName, containerName)
+	if err != nil {
+		return err
+	}
+	if ppid == 0 {
+		return nil
+	}
+
+	children, err := getProcessChildren(ppid, kclient, podName, containerName)
+	if err != nil {
+		return err
+	}
```

Comment:
Nitpick: Since these are not used until after `kill` is defined, I would move their position to be placed after `kill`.